sw/host/hsmtool/src/commands/rsa/export.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Result};
+use cryptoki::context::Pkcs11;
+use cryptoki::mechanism::Mechanism;
+use cryptoki::object::{Attribute, ObjectHandle};
+use cryptoki::session::Session;
+use rsa::{RsaPrivateKey, RsaPublicKey};
+use serde::{Deserialize, Serialize};
+use serde_annotate::Annotate;
+use std::any::Any;
+use std::path::PathBuf;
+
+use crate::commands::{BasicResult, Dispatch};
+use crate::error::HsmError;
+use crate::util::attribute::{AttrData, AttributeMap, AttributeType, KeyType, ObjectClass};
+use crate::util::helper;
+use crate::util::key::rsa::{save_private_key, save_public_key};
+use crate::util::key::KeyEncoding;
+
+#[derive(clap::Args, Debug, Serialize, Deserialize)]
+pub struct Export {
+    #[arg(long)]
+    id: Option<String>,
+    #[arg(short, long)]
+    label: Option<String>,
+    #[arg(long, help = "Export the private key")]
+    private: bool,
+    #[arg(long, help = "Wrap the exported key a wrapping key")]
+    wrap: Option<String>,
+    #[arg(short, long, value_enum, default_value = "pem")]
+    format: KeyEncoding,
+    filename: PathBuf,
+}
+
+impl Export {
+    fn export(&self, session: &Session, object: ObjectHandle) -> Result<()> {
+        let map = AttributeMap::from_object(session, object)?;
+        if self.private {
+            match map.get(&AttributeType::PrivateExponent) {
+                None => return Err(anyhow!("Key does not contain a private exponent")),
+                Some(AttrData::None) => return Err(anyhow!("Private key is not exportable")),
+                _ => {}
+            };
+            let key = RsaPrivateKey::try_from(&map)?;
+            save_private_key(&self.filename, &key, self.format)?
+        } else {
+            let key = RsaPublicKey::try_from(&map)?;
+            save_public_key(&self.filename, &key, self.format)?
+        }
+        Ok(())
+    }
+
+    fn wrap_key(&self, session: &Session, object: ObjectHandle) -> Result<()> {
+        let mut attrs = helper::search_spec(None, self.wrap.as_deref())?;
+        attrs.push(Attribute::Class(ObjectClass::SecretKey.try_into()?));
+        let mut wkey = session.find_objects(&attrs)?;
+        if wkey.is_empty() {
+            return Err(anyhow!("Found no matching wrapping key objects"));
+        } else if wkey.len() > 1 {
+            return Err(anyhow!("Found more than one matching wrapping key object"));
+        }
+        let wkey = wkey.remove(0);
+
+        // FIXME(cfrantz): Temporarily using a bogus mechanism until I figure
+        // out which mechanism should be used here.
```

Comment:
Bogus here means insecure, right? We probably want to print a big warning to the console if we hit this code path to guarantee we don't accidentally use this for anything real.

Reply:
Bogus here means "wont work and will return an error".
- RsaPkcs is not a mechanism that is allowed to wrap keys.
- The search spec for the wrapping keys looks for a key of type `SecretKey`, as RSA keys need to be wrapped with a secret key, not a public/private pair.  Trying to use any `Rsa...` mechanism with a `SecretKey` is an error.

I believe the correct mechanism here is something like `AesKeyWrap`, but I need to wait for the next release of `rust-cryptoki` for that mechanism.

Rather than leaving this half-broken, I've inserted and explicit `bail!` error and commented out the code that I know is currently broken.

Reply:
I'm glad that cryptoki checks that! Failing at runtime is fine with me. `bail!` is clearer than letting cryptoki catch the error.